### DISCUSSION TITLE
MAX7219 matrix display possible with LVGL, uDisplay and Berry ?

### DISCUSSION THREAD

**Discussion by ozett** [arendst/Tasmota#18252] [Q&A]
i want to control a simple 8x8 matrix with berry.

looks like all is in tasmota, [but complicated to build](https://github.com/arendst/Tasmota/discussions/11784) a working chain for me.
[LVGL seems to support](https://tasmota.github.io/docs/LVGL/) that MAX7219, but tasmota32-lvgl.bin must be configured to use uDisplay.
i have no clue [how to define uDisplay](https://tasmota.github.io/docs/Displays/#universal-display-driver) to make use of a 8x8 MAX7219 matrix,
it looks to me all components are available, but maybe anybody can give some hint to build a configuration-chain on tasmota for that?

simple test as working proof could be moving a single pixel on the 8x8 matrix with berry.
is this possible in general?

**Comment by sfromis** [comment]
Where did you get the idea that uDisplay and LVGL "seems to support" MAX7219?

While I'm not 100% certain, I'd not expect that combo to be supported, and if it was available, I'd have expected a [sample display descriptor file](https://github.com/arendst/Tasmota/tree/development/tasmota/displaydesc).

With Berry, you can still use the display, by firing off regular `DisplayText` commands and so on.

**Reply by ozett** [reply]
> Where did you get the idea that uDisplay and LVGL "seems to support" MAX7219?

https://docs.lvgl.io/8.1/intro/index.html#is-my-display-supported

all information about displays is so general that i was hoping that the simple matrix display 7219 could also be supported (within tasmota). to me it looks like the UDisplay definition could create the connection between tasmota hardware support and the LVGL 
(to then be able to use it with berry afterwards)

> With Berry, you can still use the display, by firing off regular `DisplayText` commands and so on.

thanks for that hint, may i try this as first step.
but i would be great to [have a control about one pixel](https://github.com/bartoszbielawski/LEDMatrixDriver/blob/master/examples/SetPixel/SetPixel.ino) (paint x,y) in the matrix itself. 
maybe an nice additon to the existing characters and text routines?

i had hoped that you could perhaps control a single pixel in the matrix with LVGL/Berry.
may i have to investigate the descriptor-file system to find a hint for Max7219 support?

**Reply by sfromis** [reply]
What I meant by my "where did you" question was primarily the uDisplay part. The point is that uDisplay was created long before LVGL support got integrated into Tasmota+Berry, and TM1637/TM1638/MAX7219 was then only supported as 7-segment displays, not really suitable for a uDisplay, which is a generalized bitmapped display driver. Matrix support for MAX7219 came later than uDisplay, as a pretty separate thing, and I'd not really see what is basically a low resolution text line display as a "natural fit" for uDisplay. Hence my expectation of such support not being there.

As usual with open source, the question may boil down to who has the combination of interest and spare time to create something which may be pretty niche when it comes to audience.

It would likely not be too difficult to add one-pixel operations to the [xdsp_19_max7219_matrix.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdsp_display/xdsp_19_max7219_matrix.ino) driver, as the used [LedControl](https://github.com/arendst/Tasmota/blob/development/lib/lib_display/LedControl/src/LedControl.h#L140) library does seem to support such.

Not knowing whether uDisplay and MAX7219 would be architecturally matching, I have no guess about how much it would take to integrate.

**Reply by ozett** [reply]
> It would likely not be too difficult to add one-pixel operations to the [xdsp_19_max7219_matrix.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdsp_display/xdsp_19_max7219_matrix.ino) driver, as the used [LedControl](https://github.com/arendst/Tasmota/blob/development/lib/lib_display/LedControl/src/LedControl.h#L140) library does seem to support such.

i looked into the driver-code and the library and it seems already in the lib: `setLed(int addr, int row, int col, boolean state)` but not available from tasmota or berry-language. i am willing to test everything, but its outside of my abilities to code something useful.

maybe somebody in this tasmota-community finds this thread and fills some lines into this niche 😄
